### ISSUE TITLE
Support for not having direct access to DbContexts

### ISSUE THREAD

**Issue by atrauzzi** [ChilliCream/hotchocolate#3043]
### Problem
I'm in the process of evaluating whether we can use Hot Chocolate to put GraphQL in front of our existing business logic layer.  This layer follows the familiar mediator pattern which means that it doesn't leak `IQueryable`.

The issue I'm having right now is that there appears to be no clear technique for getting paging or filtering information down to my business layer from Hot Chocolate.

Another problem is that HotChocolate uses multiple `DbContexts` and my handlers all use scoped `DbContexts`.

### Describe the solution you'd like
Without it having to know anything about my domain layer, I'd love for Hot Chocolate to support these kinds of setups.  I don't want to give Hot Chocolate direct access to my `DbContext` as oftentimes my business layer may be returning projections, or we have to run some logic & accounting before returning data.

Whatever the solution ends up being, it has to enable me to support projecting from my own domain models, pagination, filtering and sorting.

### Describe alternatives you've considered
 - Leaking `IQueryable` from my domain layer. :thinking: 
   - But then how do I do projections and how do I expose filtering for those projections based on the original data type?
     - Example: `User` might have `firstName` and `lastName`, but my projection might only expose `name` which is `$"{firstName} {lastName}"`.  So how would I take a filter on a field like that and apply it back to my DB model?
 - Not sure??? :laughing: 

### Additional context
Keep in mind, I'm very new to Hot Chocolate, but I can't have the default scenario it seems to be biased towards.

I'd like to add that I've tried going through the documentation, but it's been difficult to find any points where I might be able to get in and work my needs out.

:pray: Really hoping to get some worthwhile advice here, or at the very least, begin a conversation on how people can use Hot Chocolate without totally bypassing their architecture and coupling so hard to Hot Chocolate!

**Comment by michaelstaib** [comment]
Hot Chocolate itself does not know anything about data layers or DBContexts so there is actually no need to provide anything. 

The `HotChocolate.Data` package provides some filter logic for databases and also allows your to write your own filter provider. So, in order to have your very own filter logic or paging logic you need to implement providers for these.

Did you already choose a paging type (offset / cursor)?

Regarding the DBContext scoping.... you can also implement a middleware scoping your services correctly.

**Comment by PascalSenn** [comment]
Regarding pagination. It is also possible to build a custom provider for your data layer.
- [Queryable Cursor Paging](https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Core/src/Types.CursorPagination/CursorPagingProvider.cs)
- [Queryable Offset Paging](https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Core/src/Types.OffsetPagination/OffsetPagingProvider.cs)
- [MongoDB Cursor Paging](https://github.com/ChilliCream/hotchocolate/blob/develop/src/HotChocolate/MongoDb/src/Data/Paging/MongoDbCursorPagingProvider.cs)
- [MongoDb Offset Paging](https://github.com/ChilliCream/hotchocolate/blob/develop/src/HotChocolate/MongoDb/src/Data/Paging/MongoDbOffsetPagingProvider.cs)

**Comment by atrauzzi** [comment]
I'll start out with a few inline followups to highlight some difficulties I'm having.  I want to note that I feel like many of the docs I'm seeing seem to move very quickly, so I end up with lots of questions. It appears as though Hot Chocolate has a lot of extensibility in tiny pockets all over the place which as a newcomer is very overwhelming.

---

> Hot Chocolate itself does not know anything about data layers or DBContexts so there is actually no need to provide anything.

I'm not sure how this is true as the documentation I've seen talks about want to plugging a `DbContext` into Hot Chocolate and have it all pass through. Which is what prompted me to open this ticket. :slightly_smiling_face: 
---

> The HotChocolate.Data package provides some filter logic for databases and also allows your to write your own filter provider. So, in order to have your very own filter logic or paging logic you need to implement providers for these.

I guess I'm at a disadvantage because I don't know what you mean by "filter logic" here.

---

> Regarding the DBContext scoping.... you can also implement a middleware scoping your services correctly.

At the moment my context is scoped "correctly" as I register them as scoped, do you have any more details on why I would want to create a middleware?  I'm not exactly sure what the objective is or what kind of middleware we're talking about.  This seems like something I might not end up needing.

---

Anyway, in terms of the paging type I want to do, the true answer is "both".  I need to be able to support different types of paging for different parts of the graph.

@PascalSenn  - My gut tells me you might be onto the right thing here.  So I'll just ask for the sake of knowing where to look...Is there any clear way for me to simply implement a type that Hot Chocolate will resolve from DI and call a method on with all the paging information to map over to my domain layer?

I imagine this type would look like (super psuedo):

```csharp
    public class MyDataThingy : SomeHotChocolateType<IList<ReturnType>>
    {
        private readonly Mediator mediator;

        public MyDataThingy(Mediator mediator)
        {
            this.mediator = mediator;
        }

        public async Task<IList<ReturnType>> Resolve(SomeContext context)
        {
            var after = context.SomeCursorPagingMeta.After;
            var firstNameFilter = context.SomeFilterMeta["firstName"];

            return await mediator.Request<IList<ReturnType>>, MyDomainRequestType>(new MyDomainResponseType
            {
                After = after,
                FirstNameFilter = firstNameFilter,
            });
        }
    }
```

Again, this resolver would really just be mapping the pagination info to my domain layer's request type parameters for it to process.

(I might come up with a dynamic way to handle all this afterwards, but I feel like I just need to get the right mechanism.)

Overall, I'd love to just have to override **one** thing.

---

Another idea I had was if Hot Chocolate was able to offer an `Expression` tree, but the challenge there is that the types that I'd expose via GraphQL will not map 1:1 to my domain types.  So I'd need to come up with some way of mapping the `Expression` tree from the GraphQL fields, to actual filter params in my domain.  But that might be another avenue?

**Comment by PascalSenn** [comment]
So as said, HotChocolate does not need to access the DB context or doesn't even need a db context in first place.

```csharp
services.AddGraphQLServer()
    .AddQueryType<Query>();
```

Whereas query could be something like this 
```csharp
public class Query {
      public User GetUser(string name) => new User() { UserName = name};
}
```
With this you already have a function graphql server that you can query like this:
```graphql
{
     user(name: "Foo") {
          userName
     }
}
```

How you configure the GraphQL server is completely up to you. 
There are three different ways of defining a schema but you can, in your case probably want, to mix an match them
- Annotation Based 
- Code First
- Schema First
You can read about the schema basics here:

https://chillicream.com/docs/hotchocolate/defining-a-schema/

You probably want to do a mix between Code First and Annotation Based. Use Annotation based for the pure graphql related classes and code first for you domain models, where you probably don't want to use attributes

Something that could work for you is simply as that:
```csharp 
public class Query {
      public async Task<List<XYZ>> GetXYZAsync(
           [Service]Mediator mediator, 
           int? first,
           int? last,
           string? after,
           string? before)
           {
                  return await mediator.Request<IList<XZY>>, MyDomainRequestType>(new MyDomainResponseType
                  {
                      After = after,
                      Before = before,
                      First = first,
                      Last = last,
                  });
     };
}


public class XYZ Type : ObjectType<XYZ> 
{
      protected override void Configure(IObjectTypeDescriptor<XZY> descriptor)
      {
             descriptor.BindFieldsExplicitly();  // <-- If you do not do this all fields of XYZ are included by default. You can also configure this globally on the schema with ModifyOptions
             descriptor.Field(x => x.Foo);
             descriptor.Field(x => x.Bar); 
      }

}
```

```csharp
services.AddGraphQLServer()
    .AddType<XYZType>();
    .AddQueryType<Query>();
```


if you also want to use the official cursor based pagination types you have to return a Connection from you resolver:
```csharp 
public class Query {
      [GraphQLType(typeof(ConnectionType<XYZType>))]
      public async Task<Connection> GetXYZAsync(
           [Service]Mediator mediator, 
           int? first,
           int? last,
           string? after,
           string? before)
           {
                  var result =  await mediator.Request<IList<XZY>>, MyDomainRequestType>(new MyDomainResponseType
                  {
                      After = after,
                      Before = before,
                      First = first,
                      Last = last,
                  });
                  return .... // map your result into a Connection
     };
}
```

If you want to do filter in your domain, you are probably best of just defining your own filter types (in the domain) as business concernces.

(You could also use the built in filter types, but these do produce `IValueNodes` and not dotnet objects, so you probably do not want to have the HC dependency in your domain. )

```csharp 
// domain layer
public class XZYFilter 
{
     public StringFilter? UserName {get; set;}

}

public class StringFilter 
{
     public string? Eq {get; set;}
}

// graphql layer
public class Query {
      public async Task<List<XYZ>> GetXYZAsync(
           [Service]Mediator mediator, 
           int? first,
           int? last,
           string? after,
           string? before, 
           XZYFilter? where)
           {
                  return await mediator.Request<IList<XZY>>, MyDomainRequestType>(new MyDomainResponseType
                  {
                      After = after,
                      Before = before,
                      First = first,
                      Last = last,
                      Filter = where,
                  });
     };
}
```

Regarding the scoped repositories, you just have to keep one thing in mind. Each resolver of HotChocolate may execute asynchronous. 
So it could be that you dispatch requests over the mediator in parallel. Your DB context blows up if you try to do two operations at the same time. 
We try to introduce a concept of synchronisation in a later version in a couple of months. 

You see, it's completely up to you how you design your API. The `HotChocolate.Data` packages and pagination work seamlessly on IQueryable, but if you would like to not expose Queryable, you can just build it as business concernes in the domain.

**Comment by atrauzzi** [comment]
Alright, I'll try combining your points here and anything I can infer out of the documentation into a working prototype.

Some additional questions:

 - I feel like there's a `[ScopedService]` annotation that I saw somewhere.  Can I use that on my `Mediator` dependency?
 - Could I get the `IValueNodes` and then read them for values to pass to my domain request objects? Are these what HC turns GraphQL filters & pagination into?
   - I'm totally fine reading HC-specific types inside of my HC-specific `Query` class, that might even be helpful as it would let me use all of HC's functionality for filters & pagination. :slightly_smiling_face:

**Comment by atrauzzi** [comment]
A small addition, but as I'm going through this, I am hitting all kinds of issues switching over to the pooled DB context.  Currently having issues with npgsql and the nodatime extension.

Is there any way to simply have Hot Chocolate fetch things in series and respect DB context scoping?  I feel like this is blowing up basically everything downstream of it...

**Comment by atrauzzi** [comment]
I'm trying to tinker my way into getting this working, but I keep hitting error after error...  I promise, my application thus far hasn't been doing anything too fancy with DI and I've been playing very carefully within the rules.

Here's what I have right now:

```csharp
using System.Threading.Tasks;
using MyProject.Conventions.Mediator;
using MyProject.Global.Api.GraphQl.Type;
using MyProject.Global.Api.Http;
using MyProject.Global.Domain;
using MyProject.Global.Domain.Model;
using HotChocolate;
using HotChocolate.Data;

namespace MyProject.Global.Api.GraphQl
{
    public class GlobalQuery
    {
        [UseDbContext(typeof(MyProjectGlobalDbContext))]
        public async Task<Me> Me(
            [ScopedService] MyProjectGlobalDbContext dbContext,
            [Service] MyProjectGlobalContext MyProjectContext,
            [Service] Mediator mediator
        )
        {
            var response = await mediator.Request<User, Domain.Request.Me>(new Domain.Request.Me
            {
                Id = (await MyProjectContext.User).Id,
            });

            return new Me
            {
                FirstName = response.FirstName,
                LastName = response.LastName,
            };
        }
    }
}
```

As soon as I try to run my application however, I end up with the dreaded `Error while validating the service descriptor` exception, and it's a chain of basically all my dependencies, all the way to the bottom.

**Comment by PascalSenn** [comment]
- Do not use [ScopedService] for anything other than the UseDbContext. Just use [Service] for everything that comes from the DI, (transient, scoped and singleton)
- you can inject the value node of filtering like this
```csharp
public class Foo 
{
     public string Example(
          [GraphQLType(typeof(FilterInputType<Foo>))]
          IValueNode where) 
          {
                ...
          }
}

``` 
- The only need for the pooled db context is that each resolver has it's own db context and therefor parallel queries are possible. 
I wonder why you even inject DB context into the resolver when you anyway dispatch the call over mediator? 
You can also use the mediator to synchronise the dispatched requests


```

**Comment by atrauzzi** [comment]
I was only injecting the DB into the resolver because I was following the docs and didn't want to miss an annotation that HC might depend on for correctly setting things up.

Does HC create a new scope for each resolver?

**Comment by diegomaninetti** [comment]
> Does HC create a new scope for each resolver?

No, it doesn't.
But you can do it in a custom middleware. This is mine:
```
public static class GraphQLExtensions {
	private static readonly string SERVICE_SCOPE_NAME = typeof(IServiceScope).FullName ?? typeof(IServiceScope).Name;
	public static IServiceScope CreateLocalScope(this IResolverContext ctx) {
		var scope = ctx.Service<IServiceScopeFactory>().CreateScope();
		// Registering the local scope in order to close it at the end of http request.
		ctx.GetGlobalValue<HttpContext>("HttpContext").Response.RegisterForDispose(scope);
		return scope;
	}

	public static IObjectFieldDescriptor UseLocalScope(this IObjectFieldDescriptor descriptor) {
		descriptor.Use(next => async ctx => {
			try {
				ctx.SetLocalValue(SERVICE_SCOPE_NAME, ctx.CreateLocalScope());
				await next(ctx).ConfigureAwait(false);
			} finally {
				ctx.RemoveLocalValue(SERVICE_SCOPE_NAME);
			}
		});
		return descriptor;
	}

	public static T GetLocalScopedService<T>(this IResolverContext ctx) {
		return ctx.GetLocalValue<IServiceScope>(SERVICE_SCOPE_NAME).ServiceProvider.GetRequiredService<T>();
	}
}
```

Then, in your pipeline/resolver:
```
	descriptor.Field(name)
		.UseLocalScope()
		...
		.Resolver(ctx => {
			var helper = ctx.GetLocalScopedService<MyHelper>();
			/// Use your middleware scoped MyHelper class
			...
		});
```

Hope it helps ;-)

**Comment by ademchenko** [comment]
> * Leaking `IQueryable` from my domain layer. 🤔
>   
>   * But then how do I do projections and how do I expose filtering for those projections based on the original data type?
>     
>     * Example: `User` might have `firstName` and `lastName`, but my projection might only expose `name` which is `$"{firstName} {lastName}"`.  So how would I take a filter on a field like that and apply it back to my DB model?
> * Not sure??? 😆

Sometimes, or maybe even almost all times it's very important to hide the database layer from the consumer to provide them with some projection instead. From another perspective, this is quite error-prone and difficult to implement your own custom filterer even if you get the prepared expression tree from HotChocolate.

BUT: IQueryable is a pretty abstract interface that doesn't have to directly map to your database objects. That is the first point. The second point is that the HotChocolate infrastructure doesn't require you to use DataContext in services or anywhere else.

For instance, in our project, we use Postgresql and IQueryable based on [linq2db](https://linq2db.github.io/). We use neither  DataContext nor we return directly mapped IQueryable to database objects to a consumer.

That's how it is implemented:
1. We have some native service that returns IQueryable based on database objects.
2. But then we convert IQueryable to another IQueryable (the projection).
3. Since we have a pretty loaded database (millions and millions of records) it's not possible for us to materialize items before the conversion, so all conversion logic has to be applied to IQueryable. After that HotChocolate will apply filtering and paging and then we will be able to return some moderate (thanks to filtering and paging) amount of data.

```
public class Query
    {        
        [UsedImplicitly]
        public async Task<IQueryable<User>> GetUsers([Service] INativeUserTransactionService userTransactionService) => (await userTransactionService.GetAll()).Convert();
}

  public static class UsersConverter
    {
        public static IQueryable<User> Convert(this IQueryable<NativeUser> users)
        {
            return users.Select(user => new User
            {
                Id = user.Id,
                Email = user.Email,
                Phone = user.Phone != null ? System.Convert.ToString(user.Phone) : null,
                UserToken = user.Email + " " + user.Phone,
                RegistrationTime = SqlFunctions.FromUnixTimeSeconds(user.Regtime),
                FirstName = SqlFunctions.PropertyOfJsonObject(user.ProfileData, "firstName"),
                LastName = SqlFunctions.PropertyOfJsonObject(user.ProfileData, "lastName"),              
            });
        }
    }
```

I have just added a case similar to the one that you have mentioned. "UserToken" is a concatenation of the phone and email. Also, you can see that I've been able to use some other database functions like retrieving field from the JSON column ("ProfileData").

After all of that, I'm making a request with "UserToken" filter which, obviously, has to map to the database query:
```
{
  users(where: {userToken_contains: "@gmail.com 1234"})  
  {
    items
    {
      userToken
    }
  }
}
```
Omitting not important query details I can see the following trace of my database request:


```
SELECT
			user_1.email || ' ' || Cast(user_1.phone as VarChar(20)) as "UserToken",
			...
WHERE
	r."UserToken" IS NOT NULL AND r."UserToken" LIKE '%@gmail.com 1234%' ESCAPE '~'

```
You could see that the concatenated 'email phone' (which is similar to your case of 'firstName lastName') is mapped to the database query.

So, to recap: 
1. HotChocolate doesn't cause you to use DataContext.
2. IQueryable doesn't cause you to use direct database objects, you could use projected objects.
3. Neither IQueryable itself, not HotChocolate itself doesn't cause the projection 'firstName lastName' to be calculated on the materialized object. This is completely the implementation detail of the database libraries you use.

**Comment by atrauzzi** [comment]
@ademchenko - Very interesting!  Thanks for mentioning that library too BTW.

So one quick question -- if I do a projection using Entity Framework and leak that `IQueryable`, technically I should get similar to what you've described above, yes?

**Comment by ademchenko** [comment]
@atrauzzi , yes, I think the way is the same. BUT: as I mentioned before you should check if EntityFramework queryable provider supports projections you need ('firstName lastName') or if it's possible to extend its functionality.
For instance, linq2db has an awesome extending mechanism as I showed before. Actually, SqlFunctions.PropertyOfJsonObject(user.ProfileData, "lastName") is the extension I've made that provide me with the possibility to query JSON properties of an object saved in Postgresql as JSON column.
```
        [Sql.Expression("{0}->>{1}", PreferServerSide = true)]
        public static string PropertyOfJsonObject(object jsonObject, string propertyName) => throw new NotSupportedException();
```

I hope EF has some similar functionality.

**Comment by atrauzzi** [comment]
For what it's worth, based on all the information everyone has offered here, I'm happy to say that I have pretty decent support going for all my desired querying functionality.  Thank you everyone!

**Comment by kolpav** [comment]
@PascalSenn Hi, sorry to bother but this issue is best resource for HotChocolate + MediatR search query so I am commenting here.
Could you please take a look at my attempt at making HotChocolate work with MediatR and EF? I can provide full repo if needed.
I read [UseDbContext code](https://github.com/ChilliCream/hotchocolate/blob/565f24acb35a2a38c8b8153e8083e0cbd9d4248d/src/HotChocolate/Data/src/EntityFramework/Extensions/EntityFrameworkObjectFieldDescriptorExtensions.cs#L18), this [gist](https://gist.github.com/benmccallum/899a5a69a443f7157c79a57b87f6cd8b) and every related issue but I am still not sure how to do it.

```cs
// Order.cs
public class Order
{
  public int Id { get; set; }
  public string Name { get; set; }
  public int Quantity { get; set; }
}
```

```cs
// ApplicationDbContext.cs
public class ApplicationDbContext : DbContext
{
  public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options) { }
  public DbSet<Order> Orders { get; set; }
}
```

```cs
// Query.cs
public class Query
{
  [UseDbContext(typeof(ApplicationDbContext))]
  public Task<IQueryable<Order>> GetOrders([Service] IMediator mediator) =>
    mediator.Send(new GetAllOrdersQuery());
}
```

```cs
// Startup.cs
public void ConfigureServices(IServiceCollection services)
{
  services
    .AddPooledDbContextFactory<ApplicationDbContext>(options =>
      options.UseSqlite("Data Source=graphqlmediatref.db")
    )
    .AddGraphQLServer()
    .AddQueryType<Query>();

  services.AddMediatR(typeof(Startup));
}

public void Configure(IApplicationBuilder app, IWebHostEnvironment _) =>
  app
    .UseRouting()
    .UseEndpoints(endpoints => endpoints.MapGraphQL());
```

```cs
// GetAllOrdersQuery.cs
public class GetAllOrdersQuery : IRequest<IQueryable<Order>> { }

public class GetAllOrdersHandler : IRequestHandler<GetAllOrdersQuery, IQueryable<Order>>
{
  private readonly ApplicationDbContext _db;
  
  public GetAllOrdersHandler(ApplicationDbContext db) => _db = db;
  
  public Task<IQueryable<Order>> Handle(GetAllOrdersQuery request, CancellationToken cancellationToken)
    => Task.FromResult(_db.Orders.AsQueryable());
}
```

```
Unhandled exception.
System.AggregateException: Some services are not able to be constructed
...
GraphqlMediatrEf.Queries.GetAllOrders.GetAllOrdersHandler: Unable to resolve service for type 
GraphqlMediatrEf.Data.ApplicationDbContext
while attempting to activate 
GraphqlMediatrEf.Queries.GetAllOrders.GetAllOrdersHandler

```
